### PR TITLE
Fix issues with config repo partial config structure display

### LINKED ISSUE #6471: Display pipelines configured by config repo, allow linking to pipeline on dashboard
> Migrated from [Trello Card 118](https:&#x002f;&#x002f;trello.com&#x002f;c&#x002f;HeiRIZ8x)

The config repo widgets should ideally display the pipelines configured by the given CR. This is a common point in both spikes, and probably a much-wanted feature anyway.

## Related

* #6458

## And the PR

* #6360

### PR DESCRIPTION
**Pull request by marques-work**
Fixes #6471

See comment for one of the behaviors this fixes: https://github.com/gocd/gocd/issues/6471#issuecomment-506049805

  - Ensures that the resultant structure is displayed after first parse (prior to this, it just left a persistent loading message)
  - Ensures that we update the diplayed structure after forcing a reparse

Prerequisite to these fixes are significant refactoring of the ConfigRepoWidget:

  - Move a lot of the button actions/event handlers into ConfigRepoVM so that
    we are able to control/orchestrate event-driven flows
  - The ConfigRepos page widget now maintains a list of ConfigRepoVM instances
    on data fetch/refresh
  - Abstract the config repo result cache into the ConfigRepoVM as well
  - Extract utility methods into its own importable module

### CODE REVIEW

File: server/webapp/WEB-INF/rails/webpack/views/pages/config_repos/config_repo_attribute_helper.ts [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as _ from "lodash";
+import {ConfigRepo, humanizedMaterialAttributeName} from "models/config_repos/types";
+
+export function resolveHumanReadableAttributes(obj: object) {
```
Comment (kierarad) [inline server/webapp/WEB-INF/rails/webpack/views/pages/config_repos/config_repo_attribute_helper.ts]: Can you add tests for these functions?

File: server/webapp/WEB-INF/rails/webpack/views/pages/config_repos/config_repo_view_model.ts [hunk 2]
```diff
@@ -14,78 +14,126 @@
...
+        resolve(DefinedStructures.fromJSON(JSON.parse(resp.body)));
+      }, (error) => {
+        reject(error.message);
+      });
+    }).catch(rejectAsString(reject));
+  }
+
+  empty() {
+    // don't dump contents, just force a fresh set of data
+    // this.etag = stream();
```
Comment (kierarad) [inline server/webapp/WEB-INF/rails/webpack/views/pages/config_repos/config_repo_view_model.ts]: Did you mean to comment this?

File: server/webapp/WEB-INF/rails/webpack/views/pages/config_repos/spec/config_repo_view_model_spec.ts [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+  it("fetches data on the expand event", () => {
+    const vm = createVm();
+    expect(vm.results.prime).not.toHaveBeenCalled();
+
+    vm.notify("expand");
+
+    expect(vm.results.prime).toHaveBeenCalled();
+  });
+
+  it("invalidates and updates data on refresh event", () => {
```
Comment (kierarad) [inline server/webapp/WEB-INF/rails/webpack/views/pages/config_repos/spec/config_repo_view_model_spec.ts]: rename or have data update within test

File: server/webapp/WEB-INF/rails/webpack/views/pages/config_repos/spec/config_repo_view_model_spec.ts [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    const vm = createVm();
+
+    expect(vm.results.invalidate).not.toHaveBeenCalled();
+
+    vm.notify("refresh");
+
+    expect(vm.results.invalidate).toHaveBeenCalled();
+  });
+
+  it("reparseRepo() triggers reparse update invalidates the results cache", (done) => {
```
Comment (kierarad) [inline server/webapp/WEB-INF/rails/webpack/views/pages/config_repos/spec/config_repo_view_model_spec.ts]: reword?

File: server/webapp/WEB-INF/rails/webpack/views/pages/config_repos/config_repo_attribute_helper.ts (hunk 1 above)
Comment (marques-work) [reply server/webapp/WEB-INF/rails/webpack/views/pages/config_repos/config_repo_attribute_helper.ts]: @kierarad done!

File: server/webapp/WEB-INF/rails/webpack/views/pages/config_repos/config_repo_view_model.ts (hunk 2 above)
Comment (marques-work) [reply server/webapp/WEB-INF/rails/webpack/views/pages/config_repos/config_repo_view_model.ts]: That's a mistake -- it's not meant to be commented out.

File: server/webapp/WEB-INF/rails/webpack/views/pages/config_repos/spec/config_repo_view_model_spec.ts (hunk 3 above)
Comment (marques-work) [reply server/webapp/WEB-INF/rails/webpack/views/pages/config_repos/spec/config_repo_view_model_spec.ts]: will rename -- the handler is not concerned with updating data.

File: server/webapp/WEB-INF/rails/webpack/views/pages/config_repos/spec/config_repo_view_model_spec.ts (hunk 4 above)
Comment (marques-work) [reply server/webapp/WEB-INF/rails/webpack/views/pages/config_repos/spec/config_repo_view_model_spec.ts]: I was obviously ~drunk~ too tired when I wrote this.

**Comment by marques-work** [comment]
@kierarad I've rebased and made the requested changes. Merging now, and crossing fingers.

**Comment by maheshp** [comment]
Done as part of 19.6.0.

**Pull request merged**